### PR TITLE
Add yolov4-tiny.yaml and yolov4-tiny-3l.yaml

### PR DESCRIPTION
**Pull request by Duarte-Nunes**
This is a repeat PR since the old repository has to many file changes.

Hope it works now.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
The PR introduces a new CSP variant, VoVCSP, and incorporates it into two new YOLOv4-tiny model configurations.

### 📊 Key Changes
- Added a new `VoVCSP` class, a CSP (Cross Stage Partial) bottleneck variant in `experimental.py`.
- Updated `yolo.py` to recognize `VoVCSP` as a usable module in the model architectures.
- Included two new YOLOv4-tiny model configurations with different depth and width parameters in `yolov4-tiny-3l.yaml` and `yolov4-tiny.yaml`.

### 🎯 Purpose & Impact
- 🚀 **Enhanced Performance** - The `VoVCSP` bottleneck may improve model performance by enabling more efficient feature representation.
- 🧠 **New Model Variants** - The two new YOLOv4-tiny configurations expand the available options for users looking for lightweight models for object detection, potentially offering better speed and accuracy trade-offs.
- 🔁 **Versatile Usage** - Incorporating `VoVCSP` into YOLO's modules allows developers and researchers to experiment with this new structure in different contexts, fostering innovation and potentially leading to performance improvements in object detection tasks.

### CODE REVIEW

**Comment by ilem777** [comment]
@Duarte-Nunes Hello author, did you run `yolov4-tiny` successfully? I'd like to take the tiny version directly as a representative of yolov4 (because I think it's easier to perform)

**Comment by ilem777** [comment]
The training was very poor and all indicators were low.

**Comment by glenn-jocher** [comment]
@Duarte-Nunes thanks for the cleanup! Everything looks good to me. Is this ready to merge?

**Comment by Duarte-Nunes** [comment]
@glenn-jocher Altho i believe the implementation is correct i'm having trouble with the training results, simply put when i train the yolov4-tiny.yaml on a custom dataset i get bad results.

**Comment by glenn-jocher** [comment]
@Duarte-Nunes well, yolov4-tiny is probably not going to produce great results as its very small, but it should produce slightly better results than yolov3-tiny from ultralytics/yolov3:
https://github.com/ultralytics/yolov3#pretrained-checkpoints

What kind of training results are you seeing? If the results are very low (i.e. <0.10 mAP) after 300 epochs from scratch then it could indicate a bug in the architecture (either in the yaml or in the new VoVCSP module). I can take a look this week if I get some time.

**Comment by glenn-jocher** [comment]
@Duarte-Nunes hey one thing I remembered is that it looks like you have an FPN head instead of a PANet head, which should be fine, but this means the output order is reversed compared to the typical YOLOv5 model. You may want to try reversing the order of inputs into Detect, or also reversing the anchors: field from large to small instead of small to large. 

There's a section of code in autoanchor responsible for detecting reversed anchors and fixing them, but perhaps this failed for some reason.

EDIT: if the anchors are reversed this can lead to much lower mAP, i.e. 3-4X lower than normal.

**Comment by Duarte-Nunes** [comment]
@glenn-jocher Hello, i understood the step of reversing the order in Detect but i couldn't understand how to perform the reversing of the anchors step, could you elaborate?

**Comment by glenn-jocher** [comment]
@Duarte-Nunes oh sure. To reverse a model's anchor order you just reverse the row order:

```yaml
# anchors
anchors:
  - [10,14,  23,27,  37,58]  # P4/16
  - [81,82,  135,169,  344,319]  # P5/32
  
# anchors reversed
anchors:
  - [81,82,  135,169,  344,319]  # P5/32
  - [10,14,  23,27,  37,58]  # P4/16
```

**Comment by Duarte-Nunes** [comment]
@glenn-jocher Hello, apologies for the delay in answering, i have been testing the network. unfortunatly i still have not been able to get training to work even after swaping the order of the layers in detection and swapping the order of the anchors.
I'm at a loss for solutions.

**Comment by glenn-jocher** [comment]
@Duarte-Nunes I would try to train yolov3-tiny.yaml, and if you're able to train that correctly then the problem is likely in either the new yamls or the VoV module. Can you get expected results with yolov3-tiny.yaml?

**Comment by glenn-jocher** [comment]
@Duarte-Nunes by the way the training command you should use is simply:
```bash
python train.py --data coco.yaml --cfg yolov3-tiny.yaml --weights '' --batch-size 64
```

**Comment by glenn-jocher** [comment]
@Duarte-Nunes hello, any update on this? I'd like to get this merged if the bugs are sorted out. Did you try first training yolov3-tiny and comparing to yolov4-tiny with these commands?

```shell
python train.py --data coco.yaml --cfg yolov3-tiny.yaml --weights '' --batch-size 64
python train.py --data coco.yaml --cfg yolov4-tiny.yaml --weights '' --batch-size 64
```

**Comment by Duarte-Nunes** [comment]
Hey @glenn-jocher , unfortunatelly i haven´t been able to test yolov3 yet due to work but i will try to make some tests this week if thats alright and hopefully all will be resolved by the end of the week.

Thank you for your patience.

**Comment by glenn-jocher** [comment]
@Duarte-Nunes ok got it!

**Comment by zxsitu** [comment]
@Duarte-Nunes @glenn-jocher Hi guys, is the tiny running normally yet? I guess the final value might not be great, but it should work at least.

**Comment by glenn-jocher** [comment]
@ilem777 I believe the models are still not training correctly. @Duarte-Nunes can you bring us up to speed on the current status of the PR please?
